infrastructure/storage/tests/lmdb.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use serde::{Deserialize, Serialize};
+use std::{
+    net::Ipv4Addr,
+    path::{Path, PathBuf},
+    str::FromStr,
+    sync::Arc,
+    thread,
+};
+use tari_storage::lmdb_store::{db, LMDBBuilder, LMDBDatabase, LMDBError, LMDBStore};
+use tari_utilities::ExtendBytes;
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+struct User {
+    id: u64,
+    first: String,
+    last: String,
+    email: String,
+    male: bool,
+    ip: Ipv4Addr,
+}
+
+impl User {
+    fn new(csv: &str) -> Result<User, String> {
+        let vals: Vec<&str> = csv.split(",").collect();
+        if vals.len() != 6 {
+            return Err("Incomplete Record".into());
+        }
+        let id = u64::from_str(vals[0]).map_err(|e| e.to_string())?;
+        let first = vals[1].to_string();
+        let last = vals[2].to_string();
+        let email = vals[3].to_string();
+        let male = vals[4] == "Male";
+        let ip = Ipv4Addr::from_str(vals[5]).map_err(|e| e.to_string())?;
+        Ok(User {
+            id,
+            first,
+            last,
+            email,
+            male,
+            ip,
+        })
+    }
+}
+
+impl ExtendBytes for User {
```

Comment:
you should be able to derive this using the extendbytes macro

Reply:
Does the macro handle the last line?
`buf.extend_from_slice(&self.ip.to_string().as_bytes());`

Reply:
Currently I dont think it does. 
But I would add a cfg flag like it currently does with chrono to support Ipv4Addr

Reply:
Ok, so I'll leave it handrolled for now